--- a/gst/gstbufferpool.h	2014-06-01 13:20:02.986625755 +0800
+++ b/gst/gstbufferpool.h	2014-06-01 13:21:30.610629995 +0800
@@ -157,6 +157,10 @@
 
   GstFlowReturn  (*acquire_buffer) (GstBufferPool *pool, GstBuffer **buffer,
                                     GstBufferPoolAcquireParams *params);
+
+  GstFlowReturn  (*acquire_buffer_ex) (GstBufferPool *pool, GstBuffer **buffer,
+                                    GstBufferPoolAcquireParams *params, gpointer buf);
+
   GstFlowReturn  (*alloc_buffer)   (GstBufferPool *pool, GstBuffer **buffer,
                                     GstBufferPoolAcquireParams *params);
   void           (*reset_buffer)   (GstBufferPool *pool, GstBuffer *buffer);
@@ -172,6 +176,8 @@
 /* allocation */
 GstBufferPool *  gst_buffer_pool_new  (void);
 
+guint gst_buffer_pool_buffer_count(GstBufferPool *pool);
+
 /* state management */
 gboolean         gst_buffer_pool_set_active      (GstBufferPool *pool, gboolean active);
 gboolean         gst_buffer_pool_is_active       (GstBufferPool *pool);
@@ -202,6 +208,8 @@
 GstFlowReturn    gst_buffer_pool_acquire_buffer  (GstBufferPool *pool, GstBuffer **buffer,
                                                   GstBufferPoolAcquireParams *params);
 void             gst_buffer_pool_release_buffer  (GstBufferPool *pool, GstBuffer *buffer);
+GstFlowReturn    gst_buffer_pool_acquire_buffer_ex  (GstBufferPool *pool, GstBuffer **buffer,
+                                                  GstBufferPoolAcquireParams *params, gpointer buf);
 
 G_END_DECLS
 
