--- a/configure	2016-05-17 22:47:03.000000000 +0000
+++ b/configure	2016-05-20 08:58:55.851600049 +0000
@@ -69,25 +69,25 @@
     [ndk => 'ngx_devel_kit'],
-    [http_iconv => 'iconv-nginx-module', 'disabled'],
-    [http_echo => 'echo-nginx-module'],
-    [http_xss => 'xss-nginx-module'],
+    [http_iconv => 'iconv-nginx-module', 'dynamic'],
+    [http_echo => 'echo-nginx-module'],
+    [http_xss => 'xss-nginx-module', 'dynamic'],
     [http_coolkit => 'ngx_coolkit'],
-    [http_set_misc => 'set-misc-nginx-module'],
-    [http_form_input => 'form-input-nginx-module'],
-    [http_encrypted_session => 'encrypted-session-nginx-module'],
+    [http_set_misc => 'set-misc-nginx-module', 'dynamic'],
+    [http_form_input => 'form-input-nginx-module', 'dynamic'],
+    [http_encrypted_session => 'encrypted-session-nginx-module', 'dynamic'],
     [http_drizzle => 'drizzle-nginx-module', 'disabled'],
     [http_postgres => 'ngx_postgres', 'disabled'],
-    [http_srcache => 'srcache-nginx-module'],
+    [http_srcache => 'srcache-nginx-module', 'dynamic'],
     [http_lua => 'ngx_lua'],
-    [http_lua_upstream => 'ngx_lua_upstream'],
-    [http_headers_more => 'headers-more-nginx-module'],
-    [http_array_var => 'array-var-nginx-module'],
-    [http_memc => 'memc-nginx-module'],
-    [http_redis2 => 'redis2-nginx-module'],
-    [http_redis => 'redis-nginx-module'],
+    [http_lua_upstream => 'ngx_lua_upstream', 'dynamic'],
+    [http_headers_more => 'headers-more-nginx-module', 'dynamic'],
+    [http_array_var => 'array-var-nginx-module', 'dynamic'],
+    [http_memc => 'memc-nginx-module', 'dynamic'],
+    [http_redis2 => 'redis2-nginx-module', 'dynamic'],
+    [http_redis => 'redis-nginx-module', 'dynamic'],
     #[http_upstream_keepalive => 'upstream-keepalive-nginx-module'],
     #[http_auth_request => 'auth-request-nginx-module'],
-    [http_rds_json => 'rds-json-nginx-module'],
-    [http_rds_csv => 'rds-csv-nginx-module'],
+    [http_rds_json => 'rds-json-nginx-module', 'dynamic'],
+    [http_rds_csv => 'rds-csv-nginx-module', 'dynamic'],
 );
 
 my $without_resty_mods_regex;
@@ -995,7 +995,11 @@
 
         my $dir = auto_complete "../$prefix";
 
-        $opts_line .= " \\\n  --add-module=$dir";
+	if ($attr eq 'dynamic') {
+	    $opts_line .= " \\\n  --add-dynamic-module=$dir";
+	} else { 
+            $opts_line .= " \\\n  --add-module=$dir";
+	}
     }
 
     return $opts_line;
